sdk/identity/azure-identity/azure/identity/_credentials/azure_pipelines.py

```diff
@@ -0,0 +1,149 @@
+        Required.
+    :keyword str tenant_id: ID of the application's Microsoft Entra tenant. Also called its "directory" ID.
+    :keyword str client_id: The client ID of a Microsoft Entra app registration.
+    :keyword List[str] additionally_allowed_tenants: Specifies tenants in addition to the specified "tenant_id"
+        for which the credential may acquire tokens. Add the wildcard value "*" to allow the credential to
+        acquire tokens for any tenant the application can access.
+
+    .. admonition:: Example:
+
+        .. literalinclude:: ../samples/credential_creation_code_snippets.py
+            :start-after: [START create_azure_pipelines_credential]
+            :end-before: [END create_azure_pipelines_credential]
+            :language: python
+            :dedent: 4
+            :caption: Create an AzurePipelinesCredential.
+    """
+
+    def __init__(
+        self,
+        *,
+        tenant_id: str,
+        client_id: str,
+        service_connection_id: str,
+        **kwargs: Any,
+    ) -> None:
+
+        self._tenant_id = tenant_id
+        self._client_id = client_id
+        self._service_connection_id = service_connection_id
+
+        validate_tenant_id(tenant_id)
+
+        self._client_assertion_credential = ClientAssertionCredential(
+            tenant_id=tenant_id, client_id=client_id, func=self._get_oidc_token, **kwargs
+        )
+        self._pipeline = build_pipeline(**kwargs)
+
+    def get_token(
+        self,
+        *scopes: str,
+        claims: Optional[str] = None,
+        tenant_id: Optional[str] = None,
+        enable_cae: bool = False,
+        **kwargs: Any,
+    ) -> AccessToken:
+        """Request an access token for `scopes`.
+
+        This method is called automatically by Azure SDK clients.
+
+        :param str scopes: desired scopes for the access token. This method requires at least one scope.
+            For more information about scopes, see
+            https://learn.microsoft.com/entra/identity-platform/scopes-oidc.
+        :keyword str claims: additional claims required in the token, such as those returned in a resource provider's
+            claims challenge following an authorization failure.
+        :keyword str tenant_id: optional tenant to include in the token request.
+        :keyword bool enable_cae: indicates whether to enable Continuous Access Evaluation (CAE) for the requested
+            token. Defaults to False.
+
+        :return: An access token with the desired scopes.
+        :rtype: ~azure.core.credentials.AccessToken
+        :raises CredentialUnavailableError: the credential is unable to attempt authentication because it lacks
+            required data, state, or platform support
+        :raises ~azure.core.exceptions.ClientAuthenticationError: authentication failed. The error's ``message``
+            attribute gives a reason.
+        """
+        validate_env_vars()
```

Comment:
Why validate the vars in get_token rather than in ctor?

Reply:
If it were in the ctor, we couldn't support using this credential in a chainedTokenCredential (assuming we throw CredentialUnavailableException in the validate method)

Reply:
But we already have validate_tenant_id in the ctor which may throw?

Reply:
That is the same as all credentials with required parameters since the user passes those values in code - but not having the environment configured at runtime would be different.

Reply:
What's the experience in non-chained mode? Isn't raising in ctor better than raising in get_token?

Reply:
Both are runtime, so I don't think it significantly affects the experience - However, that is not true for chained mode.

Managed identity and workload identity have the same behavior.

Reply:
Right, for these SYSTEM envvars, I'd like to defer the validation to the `get_token` request so that this credential is still instantiable outside of Azure Pipelines environments (e.g. within a `ChainedTokenCredential` on a dev machine). Someone constructing this credential in non-chained mode and also trying to run this in a non-pipelines environment would hit the `CredentialUnavailableError` with the explanatory message when it is actually used, so troubleshooting should be straightforward regardless.

Reply:
The code (non-chain) is like:

cred = AzurePipelinesCredential()    <-- ctor fails here
client = Client(cred=cred)
client.call_sth()                                  <-- get_token fails here

It seems to me failure in ctor is much easier to understand for users.

For MICred and WLCred, we cannot know for sure if they work or not until get_token(), right?

Reply:
If we do validation in the constructor, then users can't put it in a chain that they can run locally unless they set all these pipeline specific envvars that they aren't ever expected to set. Do you know how we'd able to do validation in the constructor for only the non-chained case?

Reply:
add kwarg in_chain=False in the ctor?

Reply:
Eh, I feel like adding API surface area for something that amounts to whether to defer validation or not is a bit much. Users using this credential in a chain would then have to be aware of this argument, and be sure to set it, which probably  isn't the best user experience.

Reply:
Do we have data how many users are using chained credentials? (My guess is it is not a big number?)

That can help us make the trade-off.

Reply:
Not that I am aware, but I feel like we could very well be using a custom chain soon that will include AzurePipelinesCredential in the effort to move away from client/secret.

Reply:
OK. Let's make it simple. What is the behavior in other languages? Let's align with them.

Does this make sense?

Reply:
I believe all the other languages throw in GetToken (via the assertion callback)